framework/src/wifi_manager/wifi_manager.c

```diff
@@ -542,60 +621,65 @@ wifi_manager_result_e wifi_manager_set_mode(wifi_manager_mode_e mode, wifi_manag
 		softap_config.inform_new_sta_join = g_manager_callback->softap_sta_joined;
 
 		wifi_mutex_acquire(w_mutex, WIFI_UTILS_FOREVER);
-
 		g_new_sta_join = softap_config.inform_new_sta_join;
 		result = wifi_utils_start_softap(&softap_config);
-
 		if (result != WIFI_UTILS_SUCCESS) {
-			ndbg("Start softap mode fail");
-			wifi_mutex_release(w_mutex);
-			return WIFI_MANAGER_FAIL;
+			ndbg("Starting softap mode failed.");
+			goto error_with_failure;
 		}
 
 		if (start_dhcp_server() != WIFI_UTILS_SUCCESS) {
-			ndbg("start DHCP server Failed\n");
-			wifi_mutex_release(w_mutex);
-			return WIFI_MANAGER_FAIL;
+			ndbg("Starting DHCP server failed.\n");
+			goto error_with_failure;
 		}
+		wifi_mutex_release(w_mutex);
 
+		/* update wifi_manager_info */
+		wifi_mutex_acquire(w_info_mutex, WIFI_UTILS_FOREVER);
 		g_manager_info.mode = SOFTAP_MODE;
 		strncpy(g_manager_info.ssid, config->ssid, softap_config.ssid_length + 1);
 		wifi_status_set(CLIENT_DISCONNECTED);
-		wifi_mutex_release(w_mutex);
+		wifi_mutex_release(w_info_mutex);
 
 		nvdbg("Wifi mode is changed to SOFT AP");
 	}
 	// Wifi mode is changed to station
-	else if (mode == STA_MODE) {
+	else {
 		wifi_mutex_acquire(w_mutex, WIFI_UTILS_FOREVER);
-
 		result = wifi_utils_stop_softap();
 		if (result != WIFI_UTILS_SUCCESS) {
-			ndbg("Wifi stop fail");
-			wifi_mutex_release(w_mutex);
-			return WIFI_MANAGER_FAIL;
+			ndbg("Stoping softap failed.");
+			goto error_with_failure;
 		}
 
 		stop_dhcp_server();
 
 		result = wifi_utils_start_sta();
 		if (result != WIFI_UTILS_SUCCESS) {
-			ndbg("start STA fail (change STA mode fail)");
-			wifi_mutex_release(w_mutex);
-			return WIFI_MANAGER_FAIL;
+			ndbg("Starting STA failed.");
+			goto error_with_failure;
 		}
+		wifi_mutex_release(w_mutex);
 
+		/* update wifi_manager_info */
+		wifi_mutex_acquire(w_info_mutex, WIFI_UTILS_FOREVER);
 		g_manager_info.mode = STA_MODE;
 		wifi_status_set(AP_DISCONNECTED);
-		wifi_mutex_release(w_mutex);
+		wifi_mutex_release(w_info_mutex);
 
 		nvdbg("Wifi mode is changed to station");
-	} else {
-			ndbg("Invalid config!\n");
-			return WIFI_MANAGER_FAIL;
 	}
 
 	return WIFI_MANAGER_SUCCESS;
+
+error_with_failure:
+	wifi_mutex_release(w_mutex);
+	wifi_mutex_acquire(w_info_mutex, WIFI_UTILS_FOREVER);
+	g_manager_info.mode = WIFI_FAILURE;
+	wifi_mutex_release(w_info_mutex);
+	
+	return WIFI_MANAGER_FAIL;
+	
 }
 
 wifi_manager_result_e wifi_manager_get_info(wifi_manager_info_s *info)
```

Comment:
I think this code is dangerous because it's call by pointer not call by value. so application can change g_manager_info without mutex lock

Reply:
It's not possible. Please see the line below:
*info = g_manager_info;
Data copy is performed.